File: erpnext_germany/bookkeeping/report/pruefliste_monatsabschluss/pruefliste_monatsabschluss.py
# Copyright (c) 2026, ALYF GmbH and contributors
# For license information, please see license.txt

"""The checklist a month is read against before it is closed.

One line per check saying how many things it found, and under it the things
themselves, each one clickable through to the account or the voucher. A check
that found nothing still gets its line: the value of a checklist is knowing
that a question was asked, and a list that only shows problems cannot tell the
difference between a clean month and a check that never ran.
"""

import frappe
from frappe import _
from frappe.utils import cint

from erpnext_germany.bookkeeping.month_end import get_titles, run_checks
from erpnext_germany.utils.periods import get_month_range

CLEAN = "Clean"
ATTENTION = "Attention"


def execute(filters=None):
	filters = frappe._dict(filters or {})
	if not filters.company:
		frappe.throw(_("Choose a company."))

	from_date, to_date = get_period(filters)
	return get_columns(), get_data(filters, from_date, to_date)


def get_period(filters: frappe._dict):
	fiscal_year_start = frappe.get_cached_value("Fiscal Year", filters.fiscal_year, "year_start_date")
	if not fiscal_year_start:
		frappe.throw(_("Fiscal Year {0} not found").format(filters.fiscal_year))

	return get_month_range(fiscal_year_start, cint(filters.month))


def get_columns() -> list[dict]:
	return [
		{"fieldname": "label", "label": _("Check"), "fieldtype": "Data", "width": 340},
		{"fieldname": "status", "label": _("Status"), "fieldtype": "Data", "width": 110},
		{"fieldname": "count", "label": _("Findings"), "fieldtype": "Int", "width": 90},
		{"fieldname": "amount", "label": _("Amount"), "fieldtype": "Currency", "width": 150},
		{"fieldname": "detail", "label": _("Detail"), "fieldtype": "Data", "width": 420},
		{
			"fieldname": "account",
			"label": _("Account"),
			"fieldtype": "Link",
			"options": "Account",
			"width": 220,
		},
		{
			"fieldname": "voucher_no",
			"label": _("Voucher"),
			"fieldtype": "Dynamic Link",
			"options": "voucher_type",
			"width": 170,
		},
		{
			"fieldname": "voucher_type",
			"label": _("Voucher Type"),
			"fieldtype": "Data",
			"width": 140,
			"hidden": 1,
		},
	]


def get_data(filters: frappe._dict, from_date, to_date) -> list[dict]:
	results = run_checks(filters.company, from_date, to_date)
	titles = get_titles()
	only_findings = cint(filters.only_findings)

	rows = []
	for name, findings in results.items():
		if only_findings and not findings:
			continue

		rows.append(
			{
				"label": titles.get(name, name),
				"status": ATTENTION if findings else CLEAN,
				"count": len(findings),
				"amount": sum(finding.get("amount") or 0 for finding in findings),
				"indent": 0,
				"is_check": 1,
			}
		)
		rows.extend(
			{
				**finding,
				"indent": 1,
				"is_check": 0,
				"period_from_date": from_date,
				"period_to_date": to_date,
			}
			for finding in findings
		)

	return rows

File: erpnext_germany/bookkeeping/report/pruefliste_monatsabschluss/pruefliste_monatsabschluss.js
// Copyright (c) 2026, ALYF GmbH and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Pruefliste Monatsabschluss"] = {
	filters: [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
			reqd: 1,
		},
		{
			fieldname: "fiscal_year",
			label: __("Fiscal Year"),
			fieldtype: "Link",
			options: "Fiscal Year",
			default: erpnext.utils.get_fiscal_year(),
			reqd: 1,
		},
		{
			fieldname: "month",
			label: __("Month"),
			fieldtype: "Select",
			default: get_previous_month(),
			options: [
				{ value: 1, label: __("January") },
				{ value: 2, label: __("February") },
				{ value: 3, label: __("March") },
				{ value: 4, label: __("April") },
				{ value: 5, label: __("May") },
				{ value: 6, label: __("June") },
				{ value: 7, label: __("July") },
				{ value: 8, label: __("August") },
				{ value: 9, label: __("September") },
				{ value: 10, label: __("October") },
				{ value: 11, label: __("November") },
				{ value: 12, label: __("December") },
			],
			reqd: 1,
		},
		{
			fieldname: "only_findings",
			label: __("Only Checks With Findings"),
			fieldtype: "Check",
			default: 0,
		},
	],

	initial_depth: 1,

	formatter(value, row, column, data, default_formatter) {
		// The account of a finding leads to its Kontoblatt for the same period,
		// so the next question after "what is this" can be asked in one click.
		if (column.fieldname === "account" && data && data.account && !data.is_check) {
			return erpnext_germany.account_sheet.link({
				company: frappe.query_report.get_filter_value("company"),
				account: data.account,
				from_date: data.period_from_date,
				to_date: data.period_to_date,
			});
		}

		const formatted = default_formatter(value, row, column, data);

		if (column.fieldname === "status" && data && data.is_check) {
			const clean = value === "Clean";
			const label = clean ? __("Clean") : __("Attention");
			return `<span class="indicator ${clean ? "green" : "orange"}">${label}</span>`;
		}

		if (data && data.is_check) {
			return `<span style="font-weight: 600">${formatted}</span>`;
		}

		return formatted;
	},

	onload(report) {
		report.page.add_inner_button(__("File With The Company"), () => file_result(report));
	},
};

function file_result(report) {
	/* Keep the result with the company. The question an auditor asks later is
	not whether there was a checklist, but what the people who closed the month
	saw when they closed it. */
	const filters = frappe.query_report.get_filter_values();
	const [from_date, to_date] = report_period(report);

	frappe.call({
		method: "erpnext_germany.bookkeeping.month_end.archive",
		args: { company: filters.company, from_date, to_date },
		freeze: true,
		callback: (response) => {
			if (response.message) {
				frappe.show_alert({ message: __("Filed with the company."), indicator: "green" });
			}
		},
	});
}

function report_period(report) {
	/* The report already worked the period out from the fiscal year and the
	month; taking it back off a row keeps the two from disagreeing. */
	const row = (report.data || []).find((entry) => entry.period_from_date);

	return row ? [row.period_from_date, row.period_to_date] : [null, null];
}

function get_previous_month() {
	/* Return the 1-indexed number of the previous month */
	const date = new Date();
	date.setDate(0); // set to last day of previous month
	return date.getMonth() + 1; // getMonth() is 0-indexed
}
